### PR TITLE
Feature/la10/add_author_book_tag_endpoint

### PR DESCRIPTION
**Pull request by AliaksandrAnashkevich**

### CODE REVIEW

File: src/main/java/com/academia/library/controller/AuthorController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/authors")
+@RequiredArgsConstructor
+public class AuthorController {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/controller/AuthorController.java:28]: Please add swagger here

File: src/main/java/com/academia/library/controller/BookController.java [hunk 2]
```diff
@@ -37,7 +42,22 @@
             content = @Content(mediaType = "application/json", schema = @Schema(implementation = BookResponseDto.class)))
     @ApiResponse(responseCode = "404", description = "Not found", content = @Content)
     @GetMapping("{id}")
-    public BookResponseDto getBookById(@PathVariable("id") Long id){
+    public BookResponseDto getBookById(@PathVariable("id") Long id) {
         return bookService.findById(id);
     }
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/controller/BookController.java:48]: Please add swagger for all methods

File: src/main/java/com/academia/library/controller/TagController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/tags")
+@RequiredArgsConstructor
+public class TagController {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/controller/TagController.java:28]: Please add swagger configuration

File: src/main/java/com/academia/library/dto/AuthorRequestDto.java [hunk 4]
```diff
@@ -7,6 +7,6 @@
 @Getter
 public class AuthorRequestDto {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/dto/AuthorRequestDto.java]: Rename to AuthorRequest

File: src/main/java/com/academia/library/dto/AuthorResponseDto.java [hunk 5]
```diff
@@ -7,6 +7,7 @@
 @Getter
 public class AuthorResponseDto {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/dto/AuthorResponseDto.java]: Rename to AuthorResponse

File: src/main/java/com/academia/library/mapper/BookMapper.java [hunk 6]
```diff
@@ -2,16 +2,50 @@
...
+        imports = {LocalDateTime.class})
+public abstract class BookMapper {
+
+    @Autowired
+    private AuthorRepository authorRepository;
+
+    @Autowired
+    private TagRepository tagRepository;
+
+    @AfterMapping
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/mapper/BookMapper.java]: If you use AfterMapping please use it after base mapping method

File: src/main/java/com/academia/library/mapper/BookMapper.java [hunk 7]
```diff
@@ -2,16 +2,50 @@
...
+    void setTime(@MappingTarget Book book, BookRequestDto bookRequestDto) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/mapper/BookMapper.java]: setTime ? Really ?

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    @Transactional(readOnly = true)
+    public List<AuthorResponseDto> findAll() {
+        List<Author> authors = authorRepository.findAll();
+        return authors.stream()
+                .map(authorMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public AuthorResponseDto create(AuthorRequestDto authorRequestDto) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        if (authorRepository.existsByFirstNameAndLastName(authorRequestDto.getFirstName(), authorRequestDto.getLastName())) {
+            throw new AuthorAlreadyExistException(authorRequestDto.getFirstName(), authorRequestDto.getLastName());
+        }
+
+        Author author = authorMapper.toEntity(authorRequestDto);
+        return authorMapper.toDto(authorRepository.save(author));
+    }
+
+    @Override
+    public AuthorResponseDto update(Long id, AuthorRequestDto authorRequestDto) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        Author author = getAuthorOrThrowException(id);
+        author = authorMapper.updateRequestToEntity(authorRequestDto, author);
+        Author updateAuthor = authorRepository.save(author);
+        return authorMapper.toDto(updateAuthor);
+    }
+
+    @Override
+    public void delete(Long id) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java:71]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java [hunk 11]
```diff
@@ -1,43 +1,76 @@
...
     public List<BookResponseDto> findAll() {
         List<Book> books = bookRepository.findAll();
-        log.info("IN getAll - {} users found", books.size());
         return books.stream()
                 .map(bookMapper::toDto)
                 .collect(Collectors.toList());
     }
+
+    @Override
+    public BookResponseDto create(BookRequestDto bookRequestDto) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -1,43 +1,76 @@
...
+        bookValidator.validationTags(bookRequestDto.getTagsId());
+
+        Book book = bookMapper.toEntity(bookRequestDto);
+        Book createBook = bookRepository.save(book);
+
+        return bookMapper.toDto(createBook);
+    }
+
+    @Override
+    public BookResponseDto update(Long id, BookRequestDto bookRequestDto) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java [hunk 13]
```diff
@@ -1,43 +1,76 @@
...
+        bookValidator.validationTags(bookRequestDto.getTagsId());
+
+        Book book = bookMapper.mapRequestToEntity(bookRequestDto, getBookOrThrowException(id));
+        Book createBook = bookRepository.save(book);
+
+        return bookMapper.toDto(createBook);
+    }
+
+    @Override
+    public void delete(Long id) {
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/BookServiceImpl.java:71]: Please add @Transactional

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public AuthorResponseDto create(AuthorRequestDto authorRequestDto) {
+        if (authorRepository.existsByFirstNameAndLastName(authorRequestDto.getFirstName(), authorRequestDto.getLastName())) {
+            throw new AuthorAlreadyExistException(authorRequestDto.getFirstName(), authorRequestDto.getLastName());
+        }
+
+        Author author = authorMapper.toEntity(authorRequestDto);
+        return authorMapper.toDto(authorRepository.save(author));
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Please move saved author to variable

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public AuthorResponseDto update(Long id, AuthorRequestDto authorRequestDto) {
+        if (authorRepository.existsByFirstNameAndLastName(authorRequestDto.getFirstName(), authorRequestDto.getLastName())) {
+            throw new AuthorAlreadyExistException(authorRequestDto.getFirstName(), authorRequestDto.getLastName());
+        }
+
+        Author author = getAuthorOrThrowException(id);
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Use repository method here
In this case private method is redundant

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        author = authorMapper.updateRequestToEntity(authorRequestDto, author);
+        Author updateAuthor = authorRepository.save(author);
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Rename to updatedAuthor

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        return authorMapper.toDto(updateAuthor);
+    }
+
+    @Override
+    public void delete(Long id) {
+        Author author = getAuthorOrThrowException(id);
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Use Optional feature for this case

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java [hunk 18]
```diff
@@ -1,43 +1,76 @@
...
+        Book createBook = bookRepository.save(book);
+
+        return bookMapper.toDto(createBook);
+    }
+
+    @Override
+    public BookResponseDto update(Long id, BookRequestDto bookRequestDto) {
+        bookValidator.validatorAuthorAndTitleForUpdateBook(id, bookRequestDto.getAuthorId(), bookRequestDto.getTitle());
+        bookValidator.validationTags(bookRequestDto.getTagsId());
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: It doesn't look brilliant. Please think how to improve this

Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Rename to savedBook

File: src/main/java/com/academia/library/service/impl/TagServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    @Transactional
+    public void delete(Long id) {
+        Tag tag = getTagOrThrowException(id);
+        tagRepository.delete(tag);
+    }
+
+    @Transactional(readOnly = true)
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/TagServiceImpl.java]: Transactional doesn't need here

File: src/main/java/com/academia/library/service/impl/TagServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+import com.academia.library.model.Tag;
+import com.academia.library.repository.TagRepository;
+import com.academia.library.service.TagService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/service/impl/TagServiceImpl.java:17]: Follow comments in other service classes to refactor this class

File: src/main/java/com/academia/library/validator/BookValidator.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    public void validatorAuthorAndTitleForCreateBook(Long authorId, String title) {
+        if (authorRepository.existsById(authorId)) {
+            if (bookRepository.existsByTitleAndAuthorId(title, authorId)) {
+                throw new BookAlreadyExistException();
+            }
+        } else {
+            throw new AuthorNotFoundException(authorId);
+        }
+    }
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/validator/BookValidator.java]: It looks complicated a little. Please simplify this

File: src/main/java/com/academia/library/validator/BookValidator.java [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+    public void validatorAuthorAndTitleForUpdateBook(Long bookId, Long authorId, String title) {
+        if (authorRepository.existsById(authorId)) {
+            if (bookRepository.existsByTitleAndAuthorIdAndIdNot(title, authorId, bookId)) {
+                throw new BookAlreadyExistException();
+            }
+        } else {
+            throw new AuthorNotFoundException(authorId);
+        }
+    }
+
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/validator/BookValidator.java]: This method looks like method above. Please think how possible to move common logic

File: src/main/java/com/academia/library/validator/BookValidator.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    public void validationTags(List<Long> tagsId) {
+        List<Long> notExistTags = tagsId.stream()
+                .filter(Predicate.not(tagRepository::existsById))
+                .collect(Collectors.toList());
+
+        if (!notExistTags.isEmpty()) {
+            throw new TagNotFoundException(notExistTags.stream().findFirst().get());
+        }
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/validator/BookValidator.java]: It looks strange
Will discuss this method

File: src/test/java/com/academia/library/util/TestDataCreator.java [hunk 24]
```diff
@@ -1,29 +1,26 @@
...
-    public static BookRequestDto TEST_BOOK = BookRequestDto.builder()
-            .price(new BigDecimal("10.00"))
-            .title("Abstract name")
-            .createAt(LocalDateTime.now())
-            .updateAt(LocalDateTime.now())
-            .build();
+    public static String TEST_BOOK_TITLE = "Abstract name";
+
+    public static BigDecimal TEST_BOOK_PRICE = new BigDecimal("10.00");
+
```
Comment (DenisKiparizz) [inline src/test/java/com/academia/library/util/TestDataCreator.java]: In camelcase should be static final variable only. Please pay attention to basic code style recondition and "Clean Code"

**Review by DenisKiparizz — COMMENTED** [review]
Необходимо повторить транзации

File: src/main/java/com/academia/library/controller/TagController.java (hunk 3 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/TagController.java:28]: Ok, thank

File: src/main/java/com/academia/library/dto/AuthorResponseDto.java (hunk 5 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/dto/AuthorResponseDto.java]: Ok, thank

File: src/main/java/com/academia/library/mapper/BookMapper.java (hunk 6 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/mapper/BookMapper.java]: Ok, thank

File: src/main/java/com/academia/library/mapper/BookMapper.java (hunk 7 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/mapper/BookMapper.java]: Sorry, i fix this

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 8 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 9 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java (hunk 11 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 10 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java:71]: Ok, thank

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java (hunk 12 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java (hunk 13 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/BookServiceImpl.java:71]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 14 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 15 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 16 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java (hunk 17 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/AuthorServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/BookServiceImpl.java (hunk 18 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Ok, thank

Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/BookServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/TagServiceImpl.java (hunk 19 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/TagServiceImpl.java]: Ok, thank

File: src/main/java/com/academia/library/service/impl/TagServiceImpl.java (hunk 20 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/service/impl/TagServiceImpl.java:17]: Ok, thank

File: src/main/java/com/academia/library/validator/BookValidator.java (hunk 21 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/validator/BookValidator.java]: Ok, thank

File: src/main/java/com/academia/library/validator/BookValidator.java (hunk 22 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/validator/BookValidator.java]: Ok, thank

File: src/main/java/com/academia/library/validator/BookValidator.java (hunk 23 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/validator/BookValidator.java]: Ok, thank

File: src/main/java/com/academia/library/controller/AuthorController.java (hunk 1 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/AuthorController.java:28]: Ok, thank

File: src/main/java/com/academia/library/controller/BookController.java (hunk 2 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/BookController.java:48]: Ok, thank

File: src/main/java/com/academia/library/controller/TagController.java (hunk 3 above)
Comment (AliaksandrAnashkevich) [reply src/main/java/com/academia/library/controller/TagController.java:21]: Ok, thank

File: src/main/java/com/academia/library/exception/AuthorNotFoundException.java [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package com.academia.library.exception;
+
+public class AuthorNotFoundException extends RuntimeException{
+
+    private static final String AUTHOR_BOT_FOUND_MESSAGE_ERROR = "Author with id=[%s] don't exist";
```
Comment (DenisKiparizz) [inline src/main/java/com/academia/library/exception/AuthorNotFoundException.java:5]: Maybe NOT found ?
Please rename this

**Pull request merged**